ext/pcre/php_pcre.c

```diff
@@ -506,10 +506,21 @@ static int pcre_clean_cache(zval *data, void *arg)
 /* }}} */
 
 static void free_subpats_table(zend_string **subpat_names, uint32_t num_subpats) {
+	bool destroy = EG(flags) & EG_FLAGS_IN_SHUTDOWN;
```

Comment:
Bah I realised that this might not be safe.
Consider the following:
We are in a function called during shutdown (i.e. due to register_shutdown_function). Prior to the engine calling `php_call_shutdown_functions()`, the `EG_FLAGS_IN_SHUTDOWN` will already be set.
As a consequence, when during shutdown we prune an entry from the pcre cache table we will destroy the subpattern name permanently.
So if we have an array with the subpattern name in int, and then we make sure to prune the cache table, then we end up with a dangling subpattern name.

A solution to this would be to set a flag in ext/pcre's module shutdown that indicates we may destroy. This would be fairly simple and would avoid the above issue. WDYT

Reply:
Bonus: and even that is not enough.
We still may leak memory if:
1. We create an array in global scope by matching some regex and using subpattern names
2. Prune the cache, we decrement the refcount of the subpattern name string to 1
3. The global scope gets cleaned up, except the symbol table doesn't and we end up with a leak

Furthermore, if the final string release happens on hash table cleanup, we can break ZendMM because it gets cleaned up in `zend_array_destroy` with `zend_string_release_ex`.

The _proper_ solution to all of this is change how the subpattern names are cached: make the subpattern name cache per-request cache, and make it lazily populated.